S2U-21 5.1.1.17 Support for Safe Exam Browser

stetsche opened the pull request:
Adds support for safe Exam browser

https://sakaiproject.atlassian.net/browse/S2U-21

Conversion scripts: https://github.com/sakaiproject/sakai-reference/pull/187

stetsche commented on samigo/samigo-app/src/webapp/jsf/author/authorSettings.jsp:
```diff
@@ -350,17 +351,130 @@
+          <h:outputLink id="sebConfigUploadLink" rendered="#{assessmentSettings.sebConfigFileName != null}" value="/access/content#{assessmentSettings.sebConfigUploadId}" target="new_window">
+            <h:outputText value="#{assessmentSettings.sebConfigFileName}" />
+          </h:outputLink>
+          <input type="file" id="sebConfigUpload" accept=".seb" />
```
There is a label on L452

stetsche replied:
The file input has a label on L452. The Link is labelled by the filename.

mpellicer commented on config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties:
```diff
@@ -3582,10 +3582,6 @@
 
-# A list of JAR files separated by ":". Each JAR file plugin can provide one or more secure module implementation(s).
-# DEFAULT: none (null)
-# samigo.secureDeliveryPlugins=
```
Can i know why you removed this property? Maybe unused?

mpellicer replied:
Ahh I see, moved to a different section, nice.

stetsche replied:
Yeah, just a cleanup

mpellicer commented on config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties:
```diff
@@ -3686,12 +3682,20 @@
+# SAK-46501 property to select the default sorting column in assessment list view
+# The column by which we want to sort the table
+# DEFAULT: 2
+# samigo.assessmentSortingColumn=9
```
Does this belong to the feature?

mpellicer replied:
Ouch, again my bad, moved from the wrong section.

mpellicer commented on config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties:
```diff
@@ -3710,10 +3714,14 @@
-# The column by which we want to sort the table
-# DEFAULT: 2
-# samigo.assessmentSortingColumn=9
+# Safe exam Browser can be enabled globally or on a per-site basis
```
Please put the S2U-21 code before 'Safe Exam Browser'.
And put a link to the SEB website:

Ability to conduct safe exams using the 'Safe Exam Browser' open source tool
More information about SEB https://safeexambrowser.org/about_overview_en.html

mpellicer commented on config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties:
```diff
@@ -3710,10 +3714,14 @@
+# seb.enabled=always
+# seb.enabled=site
+
+# Link to download Safe Exam browser
```
Same, use the S2U-21 prefix.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages.properties:
```diff
@@ -339,6 +339,30 @@ secure_delivery_exit_pwd = Exit password (optional)
+seb_exam_keys_upload_info=For uploaded configurations the exam keys can only be set, after the assessment is \
+published. You will need to download the (altered) configuration again and insert the exam keys based on that file.
+seb_exam_keys_upload_published_info=Please insert exam keys based on the config file that you can download above. \
+The configuration has been altered, when the assessment was published.
```
Please, provide the same strings to the _ES, _CA and _EU files, I'll help with the translation.

mpellicer replied:
Add the S2U-21 code and put it at the end of the file.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AuthorMessages.properties:
```diff
@@ -423,6 +423,7 @@ sakai_assessment_rem_q=Sakai Assessment - Remove Question
 save_or_cancel_change=Save or cancel changes
 scoring=Scoring
 search_question=Search questions
+seb_exception_publish_upload=An error occured when trying to add the assessment URL to the Safe Exam Browser configuration
```
Same, provide this string to _es, _ca and _eu-

mpellicer replied:
Put this into its own section.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/DeliveryMessages.properties:
```diff
@@ -361,6 +361,10 @@ access_denied=Access Denied
 access_denied_message=You do not have permission to access this assessment. Please contact your instructor if you have any questions.
 secure_delivery_error_take_message=An error occurred while retrieving the link to take this assessment through {0}. Please contact your instructor.
 secure_delivery_error_review_message=An error occurred while retrieving the link to review this assessment through {0}.
+seb_download_configuration=Download Configuration
```
Same as above, create the strings in _es, _CA and _EU, and please, put all the strings at the end of the file into its own section.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java:
```diff
@@ -1037,26 +1063,26 @@ public void setSecureDeliveryAvailable(boolean secureDeliveryAvailable) {
+      Map.Entry entry = (Map.Entry) it.next();
+      String key = (String) entry.getKey();
+      Object o = entry.getValue();
+      if (!StringUtils.equalsAny(key,
```
Very nice!

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/DeliveryBean.java:
```diff
@@ -1320,22 +1369,19 @@ public String validate() {
+        }
+
+        if (PhaseStatus.SUCCESS.equals(secureDeliveryStatus)) {
+          results = "takeAssessment";
```
I feel takeAssessment should be a constant.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/DeliveryBean.java:
```diff
@@ -2112,20 +2164,23 @@ public String checkBeforeProceed(boolean isSubmitForGrade, boolean isFromTimer,
+          if(StringUtils.equals(secureDeliveryModuleId, SecureDeliverySeb.MODULE_NAME)) {
+            setSebSetup(true);
+          } else {
+            return "secureDeliveryError";
```
I feel "secureDeliveryError" should be a constant.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/PublishAssessmentListener.java:
```diff
@@ -285,25 +306,73 @@ private void publish(AssessmentFacade assessment,
         throw new AbortProcessingException(e);
     }
 
+	// safe Exam Browser upload
```
Identation

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java:
```diff
@@ -173,12 +177,21 @@ public void processAction(ActionEvent ae) throws AbortProcessingException
 	    assessment.updateAssessmentMetaData(SecureDeliveryServiceAPI.EXITPWD_KEY, TextFormat.convertPlaintextToFormattedTextNoHighUnicode(encryptedPassword ));
-	    
+
+        if (SecureDeliverySeb.MODULE_NAME.equals(assessmentSettings.getSecureDeliveryModule())) {
```
Identation and we need a default case.

stetsche replied:
Added default case. Didn't change indentation, because i don't wanna mix tabs and spaces.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:
```diff
@@ -338,24 +347,29 @@ public void processAction(ActionEvent ae) throws
               
               // If this is a linear access and user clicks on Show Feedback, we do not
-              // get data from db. Use delivery bean instead
+              // get data from db. UsD delivery bean instead
```
Use.

mpellicer commented on samigo/samigo-app/src/webapp/js/authoring.js:
```diff
@@ -860,3 +860,371 @@ function setAccessibilityAttributes() {
 		$(settingQuerryIdent + ", " + settingQuerryOption).first().attr("aria-describredby", helpBlockId);
 	}
 }
+class FormElement {
```
Add an space here!

mpellicer replied:
Could this logic be extracted into its own file? Something like safe.exam.browser.js ?

stetsche replied:
There is an `safeExamBrowser.js` for the delivery. we could have `settingsSafeExamBrowser.js` / `authoringSafeExamBrowser.js` and `deliverySafeExamBrowser.js`.
But I feel like other parts of `authoring.js` could use the FormElement class. There are a lot of parts in the settings where we have similar logic show/hide logic using jQuery.

mpellicer replied:
But they are not going to use them for now, I'd like to have it in a separate file for compatibility reasons. Is it possible?

stetsche replied:
sure!

mpellicer commented on samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/SecureDeliverySeb.java:
```diff
@@ -0,0 +1,330 @@
+                     + "        window.location.href = \"" + quitLink + "\";\n"
+                     + "    }\n"
+                     + "</script>\n";
+            case ASSESSMENT_REVIEW:
```
We need a default case.

mpellicer commented on samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:
```diff
@@ -401,15 +402,14 @@ private void handlePlugin( String secureDeliveryPlugin ) {
+					handleBuiltInPlugin(SecureDeliveryProctorio.class, "secureDeliveryProctorio");
+					return;
+				case "safeexambrowser":
+					handleBuiltInPlugin(SecureDeliverySeb.class, "secureDeliverySeb");
```
We need a default here.

mpellicer reviewed (changes requested):
It looks solid to me Markus @stetsche, tremendous good work!

I've requested some minor changes, please also target some codacy issues.

mpellicer commented:
We need conversion scripts for the new entities, please provide a linked PR for sakai-reference.

mpellicer commented:
@bgarciaentornos Could you also take a look?

stetsche commented:
Thanks @mpellicer. I added the requested changes and will add another patch with a few more fixes.

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AuthorMessages_eu.properties:
```diff
@@ -796,3 +796,6 @@ tag_multitag_singlequestion=Aukera hau hautatuz, galdera hau bezalakoa diren gal
 question_use_rubric=Errubrika batekin ebaluatu dezakezu galdera hau
+
+# S2U-21 Safe Exam Browser errors
+seb_exception_publish_upload=An error occured when trying to add the assessment URL to the Safe Exam Browser configuration
```
@pablosanroman Could you help us here?

PabloSanRoman replied:
Hello @mpellicer 
seb_exception_publish_upload=Errore bat gertatu da azterketaren URLa gehitu denean nabigatzaile seguruko konfiguraziora.

mpellicer replied:
Thanks Pablo!

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/DeliveryMessages_eu.properties:
```diff
@@ -570,3 +570,9 @@ savedMessage=Work is being saved.
 hideMessage=Hide Time Remaining
 showMessage=Show Time Remaining
+
+# S2U-21 Safe Exam Browser
```
@pablosanroman Could you help us here?

PabloSanRoman replied:
subMessage= Lana bidaltzeko prozesuan dago
hideMessage= Ezkutatu gainerako denbora 
showMessage= Erakutsi gainerako denbora

@mpellicer
#  S2U-21 Azterketa seguruko nabigatzailea

mpellicer replied:
Hi Pablo, I was requesting the 'seb_' strings, those strings are not related.

bgarciaentornos reviewed (changes requested):
It looks pretty mint, great work! Just found some minor things

mpellicer commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/DeliveryMessages_eu.properties:
```diff
@@ -570,3 +570,10 @@ savedMessage=Work is being saved.
+seb_download_configuration=Descargar Configuraci\u00f3n
+seb_download_seb=Descargar Safe Exam Browser
+seb_launch_seb=Ejecutar Safe Exam Browser
+seb_quit_browser=Salir de Safe Exam Browser
```
These 5 strings are the ones missing the EU translation. @PabloSanRoman

PabloSanRoman replied:
@mpellicer : 

+ begin_assessment_msg_seb_required= Azterketa honek behar du erabili “Azterketa seguruko nabigatzailea” (Safe Exam Browser). Ziurtatu Safe Exam Browser instalatuta dagoela eta egin klik exekutatu botoian.
+ seb_download_configuration=Deskargatu konfigurazioa
+ seb_download_seb=Deskargatu Safe Exam Browser
+ seb_launch_seb=Exekutatu Safe Exam Browser
+ seb_quit_browser=Atera nabigatzaile segurutik (Safe Exam Browser)

mpellicer replied:
Thanks a million!

bgarciaentornos commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages_ca.properties:
```diff
@@ -373,3 +373,26 @@ option_studentpreview=Amaga la r\u00fabrica a  l\u2019estudiant
 copyToClipboard=Copia al porta-retalls
+
+# S2U-21 Safe Exam browser support
+seb_config_mode=How do you want to configure SEB?
+seb_config_mode_client=Use client configuration
+seb_config_mode_manual=Configure Manually
+seb_config_mode_upload=Upload my own config
+seb_allow_user_quit_seb=Enable exiting from SEB
+seb_allow_reload_in_exam=Enable reload in exam
+seb_show_taskbar=Show SEB task bar
+seb_show_reload_button=Show reload button
+seb_show_time=Show time
+seb_show_keyboard_layout=Show keyboard layout
+seb_show_wifi_control=Show Wi-Fi control
+seb_allow_audio_control=Enable audio controls
+seb_allow_spell_checking=Enable spell checking
+seb_config_upload=SEB config file
+seb_config_upload_success=SEB config file was uploaded successfully.
+seb_config_upload_error=SEB config file could not be uploaded. Please make sure the configuration is a .seb and the file is not encrypted.
+seb_exam_keys=Allowed browser exam keys
+seb_exam_keys_upload_info=For uploaded configurations the exam keys can only be set, after the assessment is \
+published. You will need to download the (altered) configuration again and insert the exam keys based on that file.
+seb_exam_keys_upload_published_info=Please insert exam keys based on the config file that you can download above. \
+The configuration has been altered, when the assessment was published.
```
```suggestion
seb_config_mode=Com vols configurar Safe Exam Browser?
seb_config_mode_client=Usar la configuraci\u00f3 del client
seb_config_mode_manual=Configurar manualment
seb_config_mode_upload=C\u00e0rrega de configuraci\u00f3 pr\u00f3pia
seb_allow_user_quit_seb=Permitir eixir de Safe Exam Browser
seb_allow_reload_in_exam=Permitir rec\u00e0rrega durant l\u2019examen
seb_show_taskbar=Mostrar barra de tasques de Safe Exam Browser
seb_show_reload_button=Mostrar bot\u00f3 de rec\u00e0rrega
seb_show_time=Mostrar temps
seb_show_keyboard_layout=Mostrar la distribuci\u00f3 de teclado
seb_show_wifi_control=Mostrar control Wi-Fi
seb_allow_audio_control=Permitir controls d'\u00e0udio
seb_allow_spell_checking=Permitir la revisi\u00f3 ortogr\u00e0fica
seb_config_upload=Fitxer de configuraci\u00f3 de Safe Exam Browser
seb_config_upload_success=El fitxer de configuraci\u00f3 de Safe Exam Browser s\u2019ha carregat correctament.
seb_config_upload_error=No s\u2019ha pogut carregar el fitxer de configuraci\u00f3 de Safe Exam Browser. Per favor verific que la configuraci\u00f3 \u00e9s un fitxer .seb sense encriptar.
seb_exam_keys=Claus del navegador permeses
seb_exam_keys_upload_info=Per a configuracions pujades amb fitxer les claus de l\u2019examen solament poden establir-se quan l\u2019examen es publique.\
Descarrega la configuraci\u00f3 (alterada) de nou i insereix les claus de l\u2019examen presents en el fitxer.
seb_exam_keys_upload_published_info=Per favor, insereix les claus de l\u2019examen presents en el fitxer de configuraci\u00f3 que pots descarregar dalt. \
La configuraci\u00f3 s\u2019ha alterat quan l\u2019examen es va publicar.
```

bgarciaentornos commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages_ca.properties:
```diff
@@ -373,3 +373,26 @@ option_studentpreview=Amaga la r\u00fabrica a  l\u2019estudiant
+seb_allow_spell_checking=Permitir la revisi\u00f3 ortogr\u00e0fica
+seb_config_upload=Fitxer de configuraci\u00f3 de Safe Exam Browser
+seb_config_upload_success=El fitxer de configuraci\u00f3 de Safe Exam Browser s\u2019ha carregat correctament.
+seb_config_upload_error=No s\u2019ha pogut carregar el fitxer de configuraci\u00f3 de Safe Exam Browser. Per favor verific que la configuraci\u00f3 \u00e9s un fitxer .seb sense encriptar.
```
```suggestion
seb_config_upload_error=No s\u2019ha pogut carregar el fitxer de configuraci\u00f3 de Safe Exam Browser. Per favor verifique que la configuraci\u00f3 \u00e9s un fitxer .seb sense encriptar.
```

bgarciaentornos commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages_ca.properties:
```diff
@@ -373,3 +373,26 @@ option_studentpreview=Amaga la r\u00fabrica a  l\u2019estudiant
+seb_show_taskbar=Mostrar barra de tasques de Safe Exam Browser
+seb_show_reload_button=Mostrar bot\u00f3 de rec\u00e0rrega
+seb_show_time=Mostrar temps
+seb_show_keyboard_layout=Mostrar la distribuci\u00f3 de teclado
```
```suggestion
seb_show_keyboard_layout=Mostrar la distribuci\u00f3 de teclat
```

bgarciaentornos commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages_ca.properties:
```diff
@@ -373,3 +373,26 @@ option_studentpreview=Amaga la r\u00fabrica a  l\u2019estudiant
+seb_show_time=Mostrar temps
+seb_show_keyboard_layout=Mostrar la distribuci\u00f3 de teclado
+seb_show_wifi_control=Mostrar control Wi-Fi
+seb_allow_audio_control=Permitir controls d'\u00e0udio
```
```suggestion
seb_allow_audio_control=Permitir controls d\u2019\u00e0udio
```

stetsche commented on samigo/samigo-app/src/webapp/jsf/delivery/beginTakingAssessment.jsp:
```diff
@@ -50,6 +50,7 @@
 <!-- Secure Delivery HTML (aka Proctoring or Lockdown)-->
 <h:panelGroup rendered="#{delivery.actionString!='previewAssessment'}">
   <h:outputText value="#{delivery.secureDeliveryHTMLFragment}" escape="false" />
+  <h:outputText rendered="#{delivery.sebSetup}" value="SEB SETUP" escape="false" />
```
No, I have to remove it

stetsche replied:
Good catch

ottenhoff commented on config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties:
```diff
@@ -3694,12 +3690,20 @@
+# A list of JAR files or names for built-in modules separated by ":".
+# Each JAR file plugin can provide one or more secure module implementation(s).
+# Names for built-in modules are Proctorio and SafeExamBrowser
+# DEFAULT: SafeExamBrowser
```
default should be none

mpellicer replied:
Hey Sam, default should be SafeExamBrowser in this branch, but should be none in the community version.

ottenhoff commented on deploy/pom.xml:
```diff
@@ -660,6 +665,16 @@
+          <groupId>commons-logging</groupId>
+          <artifactId>commons-logging</artifactId>
+          <scope>compile</scope>
+        </dependency>
```
commons-logging? I thought we got rid of this one

stetsche replied:
It's included as a dependency for commons-configuration2.
https://commons.apache.org/proper/commons-configuration/dependencies.html

ottenhoff replied:
re-introducing an old dependency that was removed is prob a bad idea. if commons-configuration is already being used, just use that version ;)

mpellicer replied:
Agree with Sam, I had the same concern, let's review this Markus.

stetsche replied:
I could remove commons-logging as a dependency for commons-configuration2 without seeing side effects so far.
Currently commons-configuration (1) is used in delegatedaccess, rubrics and help and it's not in the shared libs.
As for commons-configuration2 it is used already aswell in basiclti. I would perfer to stick with the newer version, as it's used already and commons-configuration does not seem to have the best history with plist support.

ottenhoff commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/PublishAssessmentListener.java:
```diff
@@ -285,25 +306,73 @@ private void publish(AssessmentFacade assessment,
         throw new AbortProcessingException(e);
     }
 
+    // Safe Exam Browser upload
```
this code feels like it should not be here..... it should be in the SEB code

stetsche replied:
I am not sure how this could be in the SEB code, since this is not part of the delivery, but part of the settings. When an assessment is published and has an uploaded SEB config file, I need to insert the published assessment URL there and save the resource in the metadata. It would be nice if the SecureDeliveryModule spec could interact with the publishing process. But I think extending the interface would cause more trouble then this. What do you think?

ottenhoff replied:
check out the SecureDeliveryProctorio code. My questions would be:

a) Do you have to save the resource in the metadata?
b) Can you do the SEB work just-in-time (e.g., when the user wants to take the assessment) instead of doing it when the assessment is being saved?
c) Extending the interface with one or two methods would be much better than inserting product-specific code here.

mpellicer replied:
Thank you very much for the feedback Sam!

stetsche replied:
Thanks a lot Sam!

ottenhoff commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/BeginDeliveryActionListener.java:
```diff
@@ -174,6 +181,19 @@ public void processAction(ActionEvent ae) throws
+            null, Phase.ASSESSMENT_START, PhaseStatus.FAILURE, null));
+        PersistenceService persistenceService = PersistenceService.getInstance();
+        persistenceService.getSebValidationFacadeQueries()
+            .expireSebValidations(pub.getPublishedAssessmentId(), AgentFacade.getAgentString());
```
why is SEB-specific code here?

the module-specific code should be in setSecureDeliveryHTMLFragment

ottenhoff commented:
primary issues I see:

1) SEB-specific code is leaking into the Samigo code. Should try harder to keep SEB-specific code in the SEB module

2) commons-logging was removed a few versions ago

3) deploying new commons- libs could cause issues/confusion in future. prob not necessary

4) lots of altering of imports and other code. this is going to make your life more difficult to maintain branches, cherry-picks, etc

stetsche commented:
@ottenhoff Thanks for taking a look! I will investigate on Monday, how this can be addressed.

mpellicer commented:
Hey @ottenhoff , thanks for the great help reviewing this one. Did you test it by any chance? What do you feel about this feature for Sakai as as product? We're kinda excited about this!

@stetsche Let's put extra time on this work and make it more robust and optimal, as Sam said, let's get rid of old dependencies, remove non necessary cleanups and provide a lighter patch in general, that will help in a future contribution to 24.x and other 22.x instances.

ottenhoff commented:
I haven't tested it, but I wrote the Proctorio code, so I have spent a lot of time here. 

I think it's a great feature! My comments all came from experience working with the Proctorio code (e.g., lots of backports, lots of difficulty with merges, etc)

stetsche commented:
To address the the issues @ottenhoff mentioned I did the following changes
1. Seperate SEB code more from samigo
    a) SEB validations are expired within the module
    b) Added pre-delivey phases for securedelivery modules. Config edits for the happen in the publish phase within the module.
    c) What I did not move are the assessment settings and the custom begin assessment page related code.
2. Improve overall compatibility like changes on imports and other that are not needed.
3. Removed commons-logging dependency

Additionally I fixed a few more issues that I found when working on this.
 
I removed the translations for the bundle string `seb_exception_publish_upload` in favor of secure_delivery_exception_publish due to changes.
Can we add translations for `secure_delivery_exception_publish` in `AuthorMessages` please?

Thanks a lot to everyone helping with this!

ottenhoff commented:
Nice job!

stetsche commented:
Just added the conversion scripts and fixed additional codacy issues.

stetsche commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AuthorMessages_ca.properties:
```diff
@@ -827,3 +827,6 @@ restore_assessments_restore=Restaura
 restore_assessments_deleted_on=Eliminat el 
+
+# S2U-21 Secure delivery errors
+secure_delivery_exception_publish=An error related to {0} occured when trying to publish assessment.
```
Missing _ca translation

bgarciaentornos replied:
```suggestion
secure_delivery_exception_publish=S\u2019ha produ\u00eft un error  relacionat amb {0} al tractar de publicar una prova.
```

stetsche commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AuthorMessages_es.properties:
```diff
@@ -827,3 +827,6 @@ restore_assessments_restore=Restaurar
 restore_assessments_deleted_on=Fecha de borrado 
+
+# S2U-21 Secure delivery errors
+secure_delivery_exception_publish=An error related to {0} occured when trying to publish assessment.
```
Missing _es translation

bgarciaentornos replied:
```suggestion
secure_delivery_exception_publish=Se ha producido un error relacionado con {0} al intentar publicar el examen.
```

stetsche commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AuthorMessages_eu.properties:
```diff
@@ -796,3 +796,6 @@ tag_multitag_singlequestion=Aukera hau hautatuz, galdera hau bezalakoa diren gal
 question_use_rubric=Errubrika batekin ebaluatu dezakezu galdera hau
+
+# S2U-21 Secure delivery errors
+secure_delivery_exception_publish=An error related to {0} occured when trying to publish assessment.
```
Missing _eu translation

PabloSanRoman replied:
HI @bgarciaentornos 

+# S2U-21 Secure delivery errors
+secure_delivery_exception_publish= Akats bat izan da ebaluazioa argitaratzean, honekin zerikusia duena: {0}

stetsche replied:
Hi @PabloSanRoman! Awesome, thanks!

stetsche reviewed (commented):
This are the last translations, that are missing. '{0}' Could be 'Safe Exam Browser' or another secure delivery module.

bgarciaentornos commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AuthorMessages.properties:
```diff
@@ -833,3 +833,6 @@ restore_assessments_restore=Restore
 restore_assessments_deleted_on=Deleted on
+
+# S2U-21 Secure delivery errors
+secure_delivery_exception_publish=An error related to {0} occured when trying to publish assessment.
```
```suggestion
secure_delivery_exception_publish=An error related to {0} occurred when trying to publish assessment.
```

stetsche replied:
Nice catch!

stetsche commented:
I think this is ready now. Thanks to everyone who was helping with this!

mpellicer reviewed (approved):
Tremendous work!
